src/compiler/program.ts

```diff
@@ -4106,7 +4106,8 @@ export function createProgram(rootNamesOrOptions: readonly string[] | CreateProg
                 }
 
                 const isFromNodeModulesSearch = resolution.isExternalLibraryImport;
-                const isJsFile = !resolutionExtensionIsTSOrJson(resolution.extension);
+                // If this is js file source of project reference, dont treat it as js file but as d.ts
+                const isJsFile = !resolutionExtensionIsTSOrJson(resolution.extension) && !getProjectReferenceRedirectProject(resolution.resolvedFileName);

```

Comment:
I feel like I don't quite understand how this works; "is JS" seems like it should just be "has the JS extension", but this would say that if it came from a referenced project, we don't treat it as such? So, won't we misinterpret later checks regarding node module checks?

Is it the case that an equivalent fix would just be calling `getProjectReferenceRedirectProject` below in the `shouldAddFile` checks?

Reply:
for all purposes we will be adding ".d.ts" file so we are treating it as such. That happens in findSourceFile where we map the "original file name" to output "d.ts" file name